dape-tests.el

```diff
@@ -0,0 +1,230 @@
+;;; dape-tests.el --- Tests for dape.el            -*- lexical-binding: t; -*-
+
+;;; Code:
+(require 'dape)
+(require 'ert)
+
+(setq ert-batch-print-level 10)
+(setq ert-batch-print-length 15)
+
+;;; Helpers
+(defun dape--goto-line (line)
+  (goto-char (point-min))
+  (forward-line (1- line)))
+
+(defun dape--lines-with-property (property &optional value)
+  (save-excursion
+    (goto-char (point-min))
+    (let (line-numbers)
+      (while (not (eobp))
+        (let ((line-value (get-text-property (point) property)))
+          (if (or (and (not value) line-value)
+                  (and value (equal line-value value)))
+              (push (line-number-at-pos) line-numbers))
+          (forward-line 1)))
+      (nreverse line-numbers))))
+
+(defun dape--line-with-property (property &optional value)
+  (car (dape--lines-with-property property value)))
+
+(defmacro dape--with-files (file-fixtures &rest body)
```

Comment:
Make these BODY taking macros hygienic, delegate immediately to a function. Google "CALL-WITH idiom"

Reply:
Sure, is it because of stack traces. Whats your reasoning? Just curious I am allergic to "best practices"

Reply:
No, not specifically because of stack traces.  One of the benefits is that you change the implementation of
the wrapper, you don't need to find all the users of the macro and recompile them: you just recompile the implementation.  The other benefit has to do with hygiene and shadowing. By wrapping the `body` in a lambda, you guarantee that it is not affected by bindings int the macro preamble.  Another way to solve this is with `gensym` or `make-symbol`.

This is actually not unlike the problems posed by the C preprocessor.

> I am allergic to "best practices"

:-) That's a very healthy position.  Indeed always question. "best practices" is often very close to "cargo cult"

Reply:
So in your original macro I think you had something like

```
(defmacro dape--with-files (file-fixtures &rest body)
  `(let* ((temp-dir (make-temp-file "dape-tests-" t))
     ...,@body
     (remove-directory temp-dir)))
```

or something like that.  If some use of this macro a body had a `(setq temp-dir something)` you would not get an error and the code would be silently (or loudly) broken.  The alternative to `call-with` to make it hygeniec is to:

```
(defmacro dape--with-files (... &rest body)
   (let ((temp-dir-sym (gensym)))
     `(let ((,temp-dir-sym (make-temp-file...)))
          ,@body
          (remove-directory ,temp-dir-sym))))
```

But this is ugly to read and makes for funky macroexpansions.  Very common though.  The call with-idiom is much better
and has other benefits.

```
(defmacro dape--with-files (... &rest body)
   `(call-with-files (make-temp-file ...) (lambda () ,@body)))

(defun call-with-files (temp-dir fn)
   (funcall fn)
   (remove-directory temp-dir))
```